src/tuner/nccl_ofi_model.c

```diff
@@ -74,167 +66,208 @@ static const float nccl_base_lat[NCCL_NUM_ALGORITHMS][NCCL_NUM_PROTOCOLS] = {
 }
 
 double all_reduce_nvlstree_cost(struct nccl_ofi_tuner_model_dims const* dims,
 				int proto,
 				int pipe_ops,
+				const size_t chan,
 				const size_t nBytes)
 {
-	const double bandwidth = (NCCL_OFI_TUNER_INTERNODE_BW + NCCL_OFI_TUNER_INTRANODE_BW);
+	if (proto != NCCL_PROTO_SIMPLE)
+		return INFINITY;
+
+	const double ranks_per_node = ceil(dims->num_ranks * 1.0 / dims->num_nodes);
+	if (chan < ranks_per_node)
+		return INFINITY;
+
 	const double p2p_latency = nccl_nvlink_lat[NCCL_ALGO_NVLS_TREE][proto];
 	const double base_latency = nccl_base_lat[NCCL_ALGO_NVLS_TREE][proto];
-	const double num_steps_internode = 2 * log(dims->num_nodes) / log(2) - 1;
-	const double num_steps_intranode = 2 * log(dims->num_nodes) / log(2) - 1;
-	const double intranode_latency = p2p_latency * num_steps_intranode;
-	double per_tree_msg_size = 2 * nBytes;
 
-	switch (proto) {
-	case NCCL_PROTO_LL: {
-		return INFINITY;
-	}
-	case NCCL_PROTO_LL128: {
-		return INFINITY;
-	}
-	case NCCL_PROTO_SIMPLE: {
-		const double net_latency = NCCL_OFI_TUNER_COMPLETION_OVERHEAD + NCCL_OFI_TUNER_NETWORK_LATENCY;
-		const double internode_latency = (net_latency + base_latency) * num_steps_internode;
-		double latency = (intranode_latency + internode_latency);
-		return latency + (2 * per_tree_msg_size) / bandwidth;
-	}
-	default: {
-		return INFINITY;
-	}
-	}
+	const double num_trees_per_host = 2;
+	const double tree_height = floor(log(NCCL_OFI_IS_POWER_OF_TWO(dims->num_nodes) ? dims->num_nodes : dims->num_nodes - 1) / log(2)) + 1;
+
+	const double num_critical_steps_internode = /* there is a critical internode set of transfers */
+		2 * tree_height /* on the way up the tree, and again on the way back down. */
+		;
+
+	const double num_critical_steps_intranode = /* there is a critical-path intranode nvswitch process */
+		2 /* at the beginning on the leaf nodes, and at the end on the leaf nodes */
+		* ranks_per_node * num_trees_per_host
+		;
+
+	const double num_total_steps_internode = /* there is a bandwidth-consuming internode transfer */
+		dims->num_nodes /* for each host */
+		* ranks_per_node /* for each gpu on the host */
+		* 2 * tree_height /* on the way up the tree and then back down. */
+		* num_trees_per_host /* on both trees */
+		;
+
+	const double num_total_steps_intranode = /* there is a bandwidth-consuming intranode transfer */
+		dims->num_nodes /* for each host */
+		* ranks_per_node /* for each GPU */
+		* num_trees_per_host /* on both trees those ranks participate in */
+		* 2 * tree_height /* on the way up the tree and then back down. */
+		;
+
+	const double internode_bandwidth = (NCCL_OFI_TUNER_INTERNODE_BW) * dims->num_ranks;
+	const double intranode_bandwidth = (NCCL_OFI_TUNER_INTRANODE_BW) * dims->num_ranks;
+
+
+	const double internode_latency_per_step = NCCL_OFI_TUNER_NETWORK_LATENCY + NCCL_OFI_TUNER_COMPLETION_OVERHEAD;
+	const double intranode_latency_per_step = p2p_latency + base_latency;
```

Comment:
following up from line 214 comment, i suggest we add `ranks_per_node * num_trees_per_host` overhead to this latency. So, this line would be like the below:

`const double intranode_latency_per_step = (p2p_latency + base_latency) * ranks_per_node * num_trees_per_host;`

Reply:
4164ff7 fixup! tuner: better ring rank/msize binning